src/transfer-manager.ts

```diff
@@ -0,0 +1,424 @@
+/*!
+ * Copyright 2022 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Bucket, UploadOptions, UploadResponse} from './bucket';
+import {
+  DownloadCallback,
+  DownloadOptions,
+  DownloadResponse,
+  File,
+} from './file';
+import * as pLimit from 'p-limit';
+import {Metadata} from './nodejs-common';
+import * as path from 'path';
+import * as extend from 'extend';
+import {promises as fsp} from 'fs';
+
+const DEFAULT_PARALLEL_UPLOAD_LIMIT = 2;
+const DEFAULT_PARALLEL_DOWNLOAD_LIMIT = 2;
+const DEFAULT_PARALLEL_LARGE_FILE_DOWNLOAD_LIMIT = 2;
+const LARGE_FILE_SIZE_THRESHOLD = 256 * 1024 * 1024;
+const LARGE_FILE_DEFAULT_CHUNK_SIZE = 10 * 1024 * 1024;
+const EMPTY_REGEX = '(?:)';
+export interface UploadMultiOptions {
+  concurrencyLimit?: number;
+  skipIfExists?: boolean;
+  prefix?: string;
+  passthroughOptions?: UploadOptions;
+}
+
+export interface DownloadMultiOptions {
+  concurrencyLimit?: number;
+  prefix?: string;
+  stripPrefix?: string;
+  passthroughOptions?: DownloadOptions;
+}
+
+export interface LargeFileDownloadOptions {
+  concurrencyLimit?: number;
+  chunkSizeBytes?: number;
+}
+
+export interface UploadMultiCallback {
+  (err: Error | null, files?: File[], metadata?: Metadata[]): void;
+}
+
+export interface DownloadMultiCallback {
+  (err: Error | null, contents?: Buffer[]): void;
+}
+
+/**
+ * Create a TransferManager object to perform parallel transfer operations on a Cloud Storage bucket.
+ *
+ * @class
+ * @hideconstructor
+ *
+ * @param {Bucket} bucket A {@link Bucket} instance
+ */
+export class TransferManager {
+  bucket: Bucket;
+  constructor(bucket: Bucket) {
+    this.bucket = bucket;
+  }
+
+  async uploadMulti(
+    filePaths: string[],
+    options?: UploadMultiOptions
+  ): Promise<UploadResponse[]>;
+  async uploadMulti(
+    filePaths: string[],
+    callback: UploadMultiCallback
+  ): Promise<void>;
```

Comment:
This might lead to a larger discussion, but do we want to offer a `callback` interface or just `Promise`? Personally, I think most customers would want a promise-based interface and we can greatly simplify the logic for new methods going forward.

Reply:
I tend to agree with you that these days a `callback` interface seems unnecessary. The design document did call out that we would support both. Perhaps remove the callbacks and see if we get any feedback during the preview release of customers wanting them back.

Reply:
@shaffeeullah @frankyn @danielduhh How do you all feel about removing the option to provide and callback and instead only return promises from transfer manager?

Reply:
sounds good to me!

Reply:
Hey @ddelgrosso1 does `callback` provide anything other than an option for calling a function once uploadMulti is complete in addition to a promise? If not, then +1 to get feedback from customers as you called out before exposing the callback option.

Reply:
@frankyn no it does not provide anything other than what you called out. Callbacks are a bit of what I would consider "old style" Node. There usage has more or less been replaced by promises at this point.